File: api/v1/schemas/stocks.py
# -*- coding: utf-8 -*-
"""
===================================
股票数据相关模型
===================================

职责：
1. 定义股票实时行情模型
2. 定义历史 K 线数据模型
"""

from typing import Optional, List

from pydantic import BaseModel, Field


class StockQuote(BaseModel):
    """股票实时行情"""
    
    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    current_price: float = Field(..., description="当前价格")
    change: Optional[float] = Field(None, description="涨跌额")
    change_percent: Optional[float] = Field(None, description="涨跌幅 (%)")
    open: Optional[float] = Field(None, description="开盘价")
    high: Optional[float] = Field(None, description="最高价")
    low: Optional[float] = Field(None, description="最低价")
    prev_close: Optional[float] = Field(None, description="昨收价")
    volume: Optional[float] = Field(None, description="成交量（股）")
    amount: Optional[float] = Field(None, description="成交额（元）")
    update_time: Optional[str] = Field(None, description="更新时间")
    
    class Config:
        json_schema_extra = {
            "example": {
                "stock_code": "600519",
                "stock_name": "贵州茅台",
                "current_price": 1800.00,
                "change": 15.00,
                "change_percent": 0.84,
                "open": 1785.00,
                "high": 1810.00,
                "low": 1780.00,
                "prev_close": 1785.00,
                "volume": 10000000,
                "amount": 18000000000,
                "update_time": "2024-01-01T15:00:00"
            }
        }


class KLineData(BaseModel):
    """K 线数据点"""
    
    date: str = Field(..., description="日期")
    open: float = Field(..., description="开盘价")
    high: float = Field(..., description="最高价")
    low: float = Field(..., description="最低价")
    close: float = Field(..., description="收盘价")
    volume: Optional[float] = Field(None, description="成交量")
    amount: Optional[float] = Field(None, description="成交额")
    change_percent: Optional[float] = Field(None, description="涨跌幅 (%)")
    
    class Config:
        json_schema_extra = {
            "example": {
                "date": "2024-01-01",
                "open": 1785.00,
                "high": 1810.00,
                "low": 1780.00,
                "close": 1800.00,
                "volume": 10000000,
                "amount": 18000000000,
                "change_percent": 0.84
            }
        }


class ExtractItem(BaseModel):
    """单条提取结果（代码、名称、置信度）"""

    code: Optional[str] = Field(None, description="股票代码，None 表示解析失败")
    name: Optional[str] = Field(None, description="股票名称（如有）")
    confidence: str = Field("medium", description="置信度：high/medium/low")


class ExtractFromImageResponse(BaseModel):
    """图片股票代码提取响应"""

    codes: List[str] = Field(..., description="提取的股票代码（已去重，向后兼容）")
    items: List[ExtractItem] = Field(default_factory=list, description="提取结果明细（代码+名称+置信度）")
    raw_text: Optional[str] = Field(None, description="原始 LLM 响应（调试用）")


class StockHistoryResponse(BaseModel):
    """股票历史行情响应"""
    
    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    period: str = Field(..., description="K 线周期")
    data: List[KLineData] = Field(default_factory=list, description="K 线数据列表")
    
    class Config:
        json_schema_extra = {
            "example": {
                "stock_code": "600519",
                "stock_name": "贵州茅台",
                "period": "daily",
                "data": []
            }
        }


class WatchlistStockItem(BaseModel):
    """自选股管理页条目。"""

    code: str = Field(..., description="股票代码")
    name: str = Field(..., description="股票名称")
    added_at: Optional[str] = Field(None, description="添加时间")
    added_price: Optional[float] = Field(None, description="添加时价格")
    current_price: Optional[float] = Field(None, description="当前价格")
    gain_percent: Optional[float] = Field(None, description="添加后涨跌幅 (%)")
    updated_at: Optional[str] = Field(None, description="最近行情刷新时间")


class WatchlistResponse(BaseModel):
    """自选股列表响应。"""

    items: List[WatchlistStockItem] = Field(default_factory=list, description="自选股列表")


class WatchlistAddRequest(BaseModel):
    """新增自选股请求。"""

    code: str = Field(..., min_length=1, description="股票代码")
    name: Optional[str] = Field(None, description="股票名称（可选）")


class WatchlistAddResponse(BaseModel):
    """新增自选股响应。"""

    success: bool = Field(True, description="是否添加成功")
    item: WatchlistStockItem = Field(..., description="新增后的自选股条目")


class WatchlistBatchAddRequest(BaseModel):
    """批量新增自选股请求。"""

    codes: List[str] = Field(..., min_length=1, description="股票代码列表")
    name: Optional[str] = Field(None, description="股票名称，仅单只股票时生效")


class WatchlistBatchAddResultItem(BaseModel):
    """批量新增任务中的单条结果。"""

    code: str = Field(..., description="股票代码")
    status: str = Field(..., description="处理状态：success / error")
    message: str = Field(..., description="处理结果说明")
    item: Optional[WatchlistStockItem] = Field(None, description="成功新增后的自选股条目")


class WatchlistBatchTaskResponse(BaseModel):
    """批量新增任务状态响应。"""

    task_id: str = Field(..., description="任务 ID")
    status: str = Field(..., description="任务状态：running / cancelling / cancelled / completed / failed")
    total: int = Field(..., ge=0, description="总任务数")
    completed: int = Field(..., ge=0, description="已处理数量")
    results: List[WatchlistBatchAddResultItem] = Field(default_factory=list, description="逐条处理结果")
    error_message: Optional[str] = Field(None, description="任务级错误信息")
    created_at: Optional[str] = Field(None, description="任务创建时间")
    updated_at: Optional[str] = Field(None, description="任务更新时间")
    finished_at: Optional[str] = Field(None, description="任务结束时间")


class WatchlistDeleteResponse(BaseModel):
    """删除自选股响应。"""

    success: bool = Field(True, description="是否删除成功")
    code: str = Field(..., description="已删除的股票代码")

File: apps/dsa-web/src/api/stocks.ts
import apiClient from './index';

export type ExtractItem = {
  code?: string | null;
  name?: string | null;
  confidence: string;
};

export type ExtractFromImageResponse = {
  codes: string[];
  items?: ExtractItem[];
  rawText?: string;
};

export type WatchlistItem = {
  code: string;
  name: string;
  addedAt?: string | null;
  addedPrice?: number | null;
  currentPrice?: number | null;
  gainPercent?: number | null;
  updatedAt?: string | null;
};

export type WatchlistBatchResult = {
  code: string;
  status: 'success' | 'error';
  message: string;
  item?: WatchlistItem;
};

export type WatchlistBatchTask = {
  taskId: string;
  status: 'running' | 'cancelling' | 'cancelled' | 'completed' | 'failed';
  total: number;
  completed: number;
  results: WatchlistBatchResult[];
  errorMessage?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  finishedAt?: string | null;
};

export const stocksApi = {
  async extractFromImage(file: File): Promise<ExtractFromImageResponse> {
    const formData = new FormData();
    formData.append('file', file);

    const headers: { [key: string]: string | undefined } = { 'Content-Type': undefined };
    const response = await apiClient.post(
      '/api/v1/stocks/extract-from-image',
      formData,
      {
        headers,
        timeout: 60000, // Vision API can be slow; 60s
      },
    );

    const data = response.data as { codes?: string[]; items?: ExtractItem[]; raw_text?: string };
    return {
      codes: data.codes ?? [],
      items: data.items,
      rawText: data.raw_text,
    };
  },

  async parseImport(file?: File, text?: string): Promise<ExtractFromImageResponse> {
    if (file) {
      const formData = new FormData();
      formData.append('file', file);
      const headers: { [key: string]: string | undefined } = { 'Content-Type': undefined };
      const response = await apiClient.post('/api/v1/stocks/parse-import', formData, { headers });
      const data = response.data as { codes?: string[]; items?: ExtractItem[] };
      return { codes: data.codes ?? [], items: data.items };
    }
    if (text) {
      const response = await apiClient.post('/api/v1/stocks/parse-import', { text });
      const data = response.data as { codes?: string[]; items?: ExtractItem[] };
      return { codes: data.codes ?? [], items: data.items };
    }
    throw new Error('请提供文件或粘贴文本');
  },

  async getWatchlist(refresh = false): Promise<WatchlistItem[]> {
    const response = await apiClient.get<{
      items?: Array<{
        code: string;
        name: string;
        added_at?: string | null;
        added_price?: number | null;
        current_price?: number | null;
        gain_percent?: number | null;
        updated_at?: string | null;
      }>;
    }>('/api/v1/stocks/watchlist', {
      params: refresh ? { refresh: true } : undefined,
    });

    return (response.data.items ?? []).map((item) => ({
      code: item.code,
      name: item.name,
      addedAt: item.added_at,
      addedPrice: item.added_price,
      currentPrice: item.current_price,
      gainPercent: item.gain_percent,
      updatedAt: item.updated_at,
    }));
  },

  async addWatchlistStock(payload: { code: string; name?: string }): Promise<WatchlistItem> {
    const response = await apiClient.post<{
      item: {
        code: string;
        name: string;
        added_at?: string | null;
        added_price?: number | null;
        current_price?: number | null;
        gain_percent?: number | null;
        updated_at?: string | null;
      };
    }>('/api/v1/stocks/watchlist', payload);

    const item = response.data.item;
    return {
      code: item.code,
      name: item.name,
      addedAt: item.added_at,
      addedPrice: item.added_price,
      currentPrice: item.current_price,
      gainPercent: item.gain_percent,
      updatedAt: item.updated_at,
    };
  },

  async startWatchlistBatchAdd(payload: { codes: string[]; name?: string }): Promise<WatchlistBatchTask> {
    const response = await apiClient.post<{
      task_id: string;
      status: 'running' | 'cancelling' | 'cancelled' | 'completed' | 'failed';
      total: number;
      completed: number;
      results?: Array<{
        code: string;
        status: 'success' | 'error';
        message: string;
        item?: {
          code: string;
          name: string;
          added_at?: string | null;
          added_price?: number | null;
          current_price?: number | null;
          gain_percent?: number | null;
          updated_at?: string | null;
        };
      }>;
      error_message?: string | null;
      created_at?: string | null;
      updated_at?: string | null;
      finished_at?: string | null;
    }>('/api/v1/stocks/watchlist/batch', payload);

    return {
      taskId: response.data.task_id,
      status: response.data.status,
      total: response.data.total,
      completed: response.data.completed,
      results: (response.data.results ?? []).map((result) => ({
        code: result.code,
        status: result.status,
        message: result.message,
        item: result.item ? {
          code: result.item.code,
          name: result.item.name,
          addedAt: result.item.added_at,
          addedPrice: result.item.added_price,
          currentPrice: result.item.current_price,
          gainPercent: result.item.gain_percent,
          updatedAt: result.item.updated_at,
        } : undefined,
      })),
      errorMessage: response.data.error_message,
      createdAt: response.data.created_at,
      updatedAt: response.data.updated_at,
      finishedAt: response.data.finished_at,
    };
  },

  async getWatchlistBatchTask(taskId: string): Promise<WatchlistBatchTask> {
    const response = await apiClient.get<{
      task_id: string;
      status: 'running' | 'cancelling' | 'cancelled' | 'completed' | 'failed';
      total: number;
      completed: number;
      results?: Array<{
        code: string;
        status: 'success' | 'error';
        message: string;
        item?: {
          code: string;
          name: string;
          added_at?: string | null;
          added_price?: number | null;
          current_price?: number | null;
          gain_percent?: number | null;
          updated_at?: string | null;
        };
      }>;
      error_message?: string | null;
      created_at?: string | null;
      updated_at?: string | null;
      finished_at?: string | null;
    }>(`/api/v1/stocks/watchlist/batch/${encodeURIComponent(taskId)}`);

    return {
      taskId: response.data.task_id,
      status: response.data.status,
      total: response.data.total,
      completed: response.data.completed,
      results: (response.data.results ?? []).map((result) => ({
        code: result.code,
        status: result.status,
        message: result.message,
        item: result.item ? {
          code: result.item.code,
          name: result.item.name,
          addedAt: result.item.added_at,
          addedPrice: result.item.added_price,
          currentPrice: result.item.current_price,
          gainPercent: result.item.gain_percent,
          updatedAt: result.item.updated_at,
        } : undefined,
      })),
      errorMessage: response.data.error_message,
      createdAt: response.data.created_at,
      updatedAt: response.data.updated_at,
      finishedAt: response.data.finished_at,
    };
  },

  async cancelWatchlistBatchTask(taskId: string): Promise<WatchlistBatchTask> {
    const response = await apiClient.post<{
      task_id: string;
      status: 'running' | 'cancelling' | 'cancelled' | 'completed' | 'failed';
      total: number;
      completed: number;
      results?: WatchlistBatchResult[];
      error_message?: string | null;
      created_at?: string | null;
      updated_at?: string | null;
      finished_at?: string | null;
    }>(`/api/v1/stocks/watchlist/batch/${encodeURIComponent(taskId)}/cancel`);

    return {
      taskId: response.data.task_id,
      status: response.data.status,
      total: response.data.total,
      completed: response.data.completed,
      results: response.data.results ?? [],
      errorMessage: response.data.error_message,
      createdAt: response.data.created_at,
      updatedAt: response.data.updated_at,
      finishedAt: response.data.finished_at,
    };
  },

  async deleteWatchlistStock(code: string): Promise<void> {
    await apiClient.delete(`/api/v1/stocks/watchlist/${encodeURIComponent(code)}`);
  },
};
